File: zm_robot_move_relative/scripts/zm_robot_move_relative_client.py
#! /usr/bin/env python

import rospy
import actionlib
from actionlib_msgs.msg import GoalStatus
from move_base_msgs.msg import MoveBaseAction, MoveBaseGoal
from geometry_msgs.msg import PoseStamped
from tf.transformations import quaternion_from_euler
from geometry_msgs.msg import Quaternion
import math


def doneCb(state, result):
    rospy.loginfo(state)
    rospy.loginfo(result)
    rospy.loginfo(client.get_goal_status_text())


def activeCb():
    rospy.loginfo("Goal just went active")


def feedbackCb(feedback):
    pass
    #rospy.loginfo(feedback)


def goalCB(msg):
    rospy.loginfo(msg)
    goal = MoveBaseGoal()
    goal.target_pose = msg
    rospy.loginfo(goal)
    client.send_goal(goal, doneCb, activeCb, feedbackCb)


if __name__ == '__main__':
    rospy.init_node('client')
    rospy.loginfo(rospy.get_name() + " start")
    action_name = "zm_robot_move_relative"
    client = actionlib.SimpleActionClient(action_name, MoveBaseAction)
    rospy.loginfo("wait_for_server: " + action_name)
    client.wait_for_server()
    rospy.loginfo("get server: " + action_name)
    rviz_goal_Subscriber = rospy.Subscriber("/move_base_simple/goal",
                                            PoseStamped, goalCB)
    r = rospy.Rate(10)
    while not rospy.is_shutdown():
       rospy.loginfo("input goal: ")
       goal = MoveBaseGoal()
       goal.target_pose.header.frame_id = raw_input("frame_id:") or "base_link"
       goal.target_pose.pose.position.x = float(raw_input("x (meter): "))
       goal.target_pose.pose.position.y = float(raw_input("y (meter): "))
       goal.target_pose.pose.orientation = Quaternion(*quaternion_from_euler(0, 0, math.radians(float(raw_input("theta (degree): ")))))
       rospy.loginfo(goal)
       client.send_goal(goal, doneCb, activeCb, feedbackCb)

File: zm_robot_move_relative/src/zm_robot_move_relative_client.cpp
#include <iostream>
#include <actionlib/client/simple_action_client.h>
#include <move_base_msgs/MoveBaseAction.h>
#include <move_base_msgs/MoveBaseGoal.h>
#include <geometry_msgs/PoseStamped.h>
#include <geometry_msgs/Quaternion.h>
#include <tf/transform_broadcaster.h>

int main(int argc, char** argv)
{
  
  ros::init(argc, argv, "client");
  //ROS_INFO(ros::this_node::getName() + " start");
 
  ros::NodeHandle n;

  // create the action client
  // true causes the client to spin its own thread
  actionlib::SimpleActionClient<move_base_msgs::MoveBaseAction> ac("zm_robot_move_relative", true);
  ac.waitForServer(); //will wait for infinite time

  //std::string Node_name = ros::this_node::getName();

  //ROS_INFO("get server : " + Node_name);
  
  ros::Rate loop_rate(10);
  while(n.ok())
  {
   
    std::string frame_id;
    std::string x, y, theta;

    std::cout << "Please enter frame:";
    std::getline(std::cin, frame_id);
    std::cout << "Please enter x position:";
    std::getline(std::cin, x);
    std::cout << "Please enter y position:";
    std::getline(std::cin, y);
    std::cout << "Please enter theta position:";
    std::getline(std::cin, theta);

    move_base_msgs::MoveBaseGoal goal;

    goal.target_pose.header.frame_id = frame_id;
    goal.target_pose.header.stamp = ros::Time::now();

    goal.target_pose.pose.position.x = stod(x);
    goal.target_pose.pose.position.y = stod(y);
    goal.target_pose.pose.orientation = tf::createQuaternionMsgFromYaw(stod(theta));

    //ROS_INFO(goal.c_str());

    ac.sendGoal(goal);

    ac.waitForResult();

    if(ac.getState() == actionlib::SimpleClientGoalState::SUCCEEDED)
      ROS_INFO("Hooray, the base move.");
    else
      ROS_INFO("The base failed to move.");
  }

  return 0;
}
